<!DOCTYPE html
 PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
     "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<!--
************************************************************************
* Project: Packaging the Nonnegative Matrix Factorization Algorithm
* Paper: Position-Dependent Motif Characterization Using Nonnegative 
*        Matrix Factorization 
* Publication: Bioinformatics (Oxford, England) (13 October 2008)
* Date: December 2008
* Authors: Joel Graber, Lucie Hutchins,Sean M M Murphy, Priyam Singh
* Organization : The Jackson Laboratory, Bar-Harbor Maine, USA.
*************************************************************************
-->
<head><title>Nonnegative Matrix Factorization Pipeline</title>
  <link  href="style.css" rel="stylesheet" type="text/css" />
</head>
<body>
      <div id="header"><h1>Non-negative Matrix Factorization Pipeline</h1></div>
      <div id="navband">
           <ul><li><a href="index.html">Readme</a></li>
               <li><a href="howtorun.html">How To Run</a> </li>
               <li><a href="docs/nmfOutputFiles.txt">Results Files</a></li>
               <li><a href="docs/nmfPipelineArguments.txt">Pipeline Arguments</a></li>
      </div>
     <div id="maindiv"><div id="lp">
 <h1>How to run the pipeline</h1>

 <ol><li><h2>Step 0</h2>
    The input file must be fasta-formatted, and all sequences are assumed to be the same length, and aligned on a common feature of some sort.  

<p>The common length of the sequences must be known in advance and passed to the pipeline program with the "-L" option.  The alignment point can be set if desired with the "-z" option.  If it's not set, the beginning of all sequences is taken as position 0.  (For instance in the example file that you have, all sequences are 200 nt up- and downstream of the polyA site, so I would include the arguments "-L" 400 and "-z" 200.  This would result in one count window boundary set at position 200 of the sequences (with this called position 0) and all other windows set up accordingly.</p>

<p>The "-a" argument typically will be set to 1, the beginning of the process (the call to fa2stats), however if the files generated by these programs have already been created, the process can be started at later steps.  For instance, the first step (fa2stats) only needs to be run once on a given fasta-formatted sequence file.  If the same sequence file is analyzed again with different values for "-w" or "-k" there is no reason to start at "-a" 1.  Similarly, if the choices for "-w" and "-k" are unchanged and a new value of "-r" (the number of patterns), then the first three steps are redundant, so "-a" 4 would be appropriate. </p> 

<p>What is critical to understand is that the pipeline program assumes not only that the earlier programs will have been run, but also that the naming of the output files from these runs can be predicted from the parameters of the current run.  If this condition is not met, it will likely fail, and in the current version of the programs, it does not fail gracefully.</p>

    </li><li> <h2>Step 1</h2> Store your input fasta-formatted file (myinputfile.fa ) under data directory
         $NMF_PATH/data/myinputfile.fa
 <br/> Note: Input myinputfile.fa format: A fasta of m equal size sequences .<br/>
  For example:
     <br/>>1      786
  <br/>
AATGCTTGTTTATATGTTTGTTTCTGCTAACTTGTCTTGTTGTTCTTGAAGCTGACGTGTGTTT
<br/>
TGTCTTTCAGGAATTCACTGATCATCTGGCCAAGACTCACACCAGGGTCTCCGTTCAGAGGACT
<br/>
CAGGGAGCTCTTCCTCAGGGAACCTCCTAAATTTTGTACAACGGATTGAAATAAAAGTCCAGAA
<br/>
GTAAAACCATCTTTGAGTGTATGGTCTGTTCTTGGGCGTTAAAGTTGTGTGAAAGGTGCAAAAT
<br/>
AATATCCTGATATTACTGGCTTTCCCATGGTTCCTTAATCTTTTTAAAGTTTGTGATTCTGATT
<br/>
ACATTAGATGCTGGAAAGTGTTGGAAAATAAGCATTAATAATGGTCCTTGCTTGAATTTGTTTA
<br/>
ATGCTGCGTCCCAATT
 <br/>>2      886
<br/>
TGTCTTTCAGGAATTCACTGATCATCTGGCCAAGACTCACACCAGGGTCTCCGTTCAGAGGACT
<br/>
CAGGGAGCTCTTCCTCAGGGAACCTCCTAAATTTTGTACAACGGATTGAAATAAAAGTCCAGAA
<br/>
GTAAAACCATCTTTGAGTGTATGGTCTGTTCTTGGGCGTTAAAGTTGTGTGAAAGGTGCAAAAT
<br/>
AATATCCTGATATTACTGGCTTTCCCATGGTTCCTTAATCTTTTTAAAGTTTGTGATTCTGATT
<br/>
ACATTAGATGCTGGAAAGTGTTGGAAAATAAGCATTAATAATGGTCCTTGCTTGAATTTGTTTA
 <br/>
AATGCTTGTTTATATGTTTGTTTCTGCTAACTTGTCTTGTTGTTCTTGAAGCTGACGTGTGTTT
<br/>
ATGCTGCGTCCCAATT
     <br/>....
 <br/>In this example we have a fasta file containing genome aligned sequences of length=400bp
 <br/>word_len=400 bp<br/>
 </li>
 <br/>
 <li><h2>Step 2</h2> <b>Check that third-party libraries are installed</b>
      <br/>Change directory to <b>sh_script</b> and run the <b>install</b> script from the command line : ./install
 <!-- b. Run the setup script from the cammand line
    %> ./setup -->
       <br/> If the install script is a success and displays : "gsl and gnuplot are installed.You can run the pipeline"
            then <b>skip steps 3 and 4</b>
  </li><br/>
   <li><h2>Step 3</h2> <b>Install third-party libraries </b>
      <br/>You can use our gsl package under $NMF_PATH/bin/libpck/gsl-1.8.tar.gz to install the gsl libraries. Make sure that the gsl package
        is installed under  /usr/local/include/  ( /usr/local/include/gsl)
      <br/> You can use our gnuplot package under $NMF_PATH/bin/libpck/gnuplot-4.2.4.tar.gz to install gnuplot. Make sure that gnuplot is in your path. You can invoke gnuplot from the command line.
      <br/>After installing third-party programs and libraries, re-run the install script to verify that the nmf supporting programs can find the libraries
       <br/>Change directory to <b>sh_script</b> and run the <b>install</b> script from the command line :  ./install
       <br/> If the install script is a success and displays : "gsl and gnuplot are installed.You can run the pipeline"
            then <b>skip step 4</b>
   </li>
   <br/>
   <li><h2>Step 4</h2> <b>Run the setup Script </b>
      <br/>Change directory to <b>sh_script</b> and run the <b>setup</b> script from the command line : ./setup
      <br/>Change directory to <b>sh_script</b> and run the <b>install</b> script from the command line :  ./install
       <br/> If the install script is a success and displays : "gsl and gnuplot are installed.You can run the pipeline"
            then <b>Proceed with step 5, otherwise check the error logs under logs/ fix the problem and try to re-run the setup and the install scrpits</b>
   </li>
   <br/>
   <li><h2>Step 5</h2> <b>Running the pipeline </b>
      <!--<br/>Change directory to $NMF_PATH/pipeline/   -->
      <br/>Open the Configuration file($NMF_PATH/sh_script/pipeline_arg_config) and setup the pipeline arguments:
        <ul><li>Set the  Absolute path to nmf package($NMF_PATH)</li>
            <li> Modify the input file name ($INPUT_FILE)</li>
            <li>Set the word length ($WORD_LEN)</li>
            <li>Set the output prefix ($OUTPUT_PREFIX)</li>
             <li>Set the start action ($A)</li>
           <li>Set optional arguments as needed (only default values of optional arguments are used at the moment) </li>
      <li>Change directory to $NMF_PATH/sh_script and run the pipeline script (run_pipeline) on the command line: ./run_pipeline</li>
      <li>You can also run the pipeline as a background process using the command :<b> nohup ./run_pipeline & </b></li></ul>
   </li>
   <li id="nc"><h2>Step 6</h2> <b>Pipeline result files </b>
      <br/>The generated result files are under $NMF_PATH/pipeline/result/ . Click on the <a  href="docs/nmfOutputFiles.txt"> "RESULT FILES" </a>link ( $NMF_PATH/docs/nmfOutputFiles.txt )
        on the nmf main page to read the description of each file.
      <br/>The generated logs, errors files can be found under $NMF_PATH/logs/
   </li>
   </ol>
</div>
 <div id="rnav">
       <div class="cheader">Directories</div>
       <ul><li><a href="index.html">home</a></li>
              <li><a href="sh_script/">Shell scripts</a></li>
              <li><a href="pipeline">pipeline</a></li>
              <li><a href="data">Data</a></li><li><a href="docs">Docs</a></li>
              <li><a href="bin/">Bin</a></li>
                <li><a href="src">src</a>
               </li></ul>
  </div>
 </div>
</body>
</html>

